### ISSUE TITLE
ftflow filtering not working

### ISSUE THREAD

**Issue by marcopp2018** [ntop/PF_RING#398]
Hi, 

I have used the ftflow.c example and uncomment the following commands:
     pfring_ft_load_configuration(ft, "rules.conf");

and following rules.conf file:

[shunt]
default = 10
tcp = 15
udp = 2
HTTP = 10
SSL = 2

[filter]
YouTube = discard
Netflix = discard
DNS = discard

and also tried the following:

   pfring_ft_set_filter_protocol_by_name(ft, "MDNS", PFRING_FT_ACTION_DISCARD);
   pfring_ft_set_filter_protocol_by_name(ft, "UPnP", PFRING_FT_ACTION_DISCARD);

In both tests the result were not consistent, for instance MDNS wasn't discard, or in the above 
YouTube wasn't discard. 
is there a way to debug this?
Thanks.

**Comment by cardigliano** [comment]
Do you mean that the action returned by pfring_ft_process() is other than PFRING_FT_ACTION_DISCARD for MDNS/UPnP packets?
Do you see MDNS/UPnP as "l7" metadata in records logged by processFlow()?

**Comment by marcopp2018** [comment]
using the -v flag I can see that the packets action is discard:

`[Packet] [discard] [04:0C:CE:DA:EF:48 -> 01:00:5E:00:00:FB] [IPv4][10.0.0.13:5353 -> 224.0.0.251:5353] [l3_proto=UDP][hash=3925879531] [caplen=347][len=347][eth_offset=0][l3_offset=14][l4_offset=34][payload_offset=42]
`

The same flow with discard packets appears in the metadata records:

`[Flow] l7: MDNS, category: 14, srcIp: 10.0.0.13, dstIp: 224.0.0.251, srcPort: 5353, dstPort: 5353, protocol: 17, tcpFlags: 0x00, c2s: { Packets: 14, Bytes: 2869, First: 1535301688.565666, Last: 1535301696.183762 }, s2c: { Packets: 0, Bytes: 0, First: 0.0, Last: 0.0 }
`

**Comment by cardigliano** [comment]
@marcopp2018, it seems that the filtering engine is working, are you still experiencing issues?

**Comment by marcopp2018** [comment]
Yes,
I cloned the project again, and compile kernel and libraries.
Uncomment the MDNS filtering command ("pfring_ft_set_filter_protocol_by_name(ft, "MDNS", PFRING_FT_ACTION_DISCARD);")
make clean; make

Got the following when running ./ftflow -i eth0 -7 -v:

[Packet] **[discard]** [C0:F2:FB:77:A8:98 -> 01:00:5E:00:00:FB] [IPv4][10.0.0.11:5353 -> 224.0.0.251:5353] [l3_proto=UDP][hash=3925879529] [caplen=105][len=105][eth_offset=0][l3_offset=14][l4_offset=34][payload_offset=42]
[Packet] **[discard]** [C0:F2:FB:77:A8:98 -> 01:00:5E:00:00:FB] [IPv4][10.0.0.11:5353 -> 224.0.0.251:5353] [l3_proto=UDP][hash=3925879529] [caplen=105][len=105][eth_offset=0][l3_offset=14][l4_offset=34][payload_offset=42]


[Flow] l7: MDNS, category: 14, srcIp: 10.0.0.11, dstIp: 224.0.0.251, srcPort: 5353, dstPort: 5353, protocol: 17, tcpFlags: 0x00, c2s: { Packets: 2, Bytes: 210, First: 1535357663.608317, Last: 1535357664.586604 }, s2c: { Packets: 0, Bytes: 0, First: 0.0, Last: 0.0 }

Packets are being discard but the flow appears in the flow summary.
Am I missing something? Do I need to use an additional function in order to us the  filtering engine?

BTW - After the fix for layer 7 yesterday the application has Segmentation fault when not using -7 flag

**Comment by cardigliano** [comment]
@marcopp2018 do you want flows to be discarded as well? Please note that the library needs to keep flows anyway, does it work for you if you are able to check the flow 'action' when exporting a flow, and skip it in the summary? Is this what you are looking for?
As of the segfault, I am not able to reproduce it, could you provide more details for reproducing it?

**Comment by marcopp2018** [comment]
Ok I thing I got it. 
in order to implement for instance L7 firewall there is a need to implement function that checks the action output and forward only packets that weren't discard. 

> Please note that the library needs to keep flows anyway, does it work for you if you are able to check the flow 'action' when exporting a flow, and skip it in the summary? Is this what you are looking for?

Yes, is there a way to do it? action per flow rather than per packet

In regards to the  segmentation fault I used MDNS  filter when layer 7 isn't enabled, my mistake.
